### PR TITLE
[WIP] :Added support for Avro and JSON records

### PR DESCRIPTION
**Pull request by anmolbansal144**
#### Pull Request Description
- Added following formatters :
  1. ByteArrayFormatter
  2. StringFormatter
  3. AvroFormatter
  4. Json Formatter
- Added integeration test for Json and csv

### CODE REVIEW

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 1]
```diff
@@ -107,6 +107,14 @@
         + "the Connector makes to ingest records into KustoDB.";
     private static final String KUSTO_SINK_RETRY_BACKOFF_TIME_MS_DISPLAY = "Retry BackOff Time";
 
+    private static final String AVRO_CODEC_CONFIG = "avro.codec";
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:110]: I guess this is not required. Let's just write the avro data directly.

File: pom.xml [hunk 2]
```diff
@@ -140,6 +146,92 @@
             <version>1.9.5</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.kafka</groupId>
```
[the remark this answers is not in the document]
Comment (fahadhasher) [reply pom.xml:150]: Or let's remove the integration tests for now and give them as a separate PR later.
This way PR will be cleaner.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java [hunk 3]
```diff
@@ -102,31 +103,13 @@ String getFilePath(@Nullable Long offset) {
...
-        if (value == null) {
+        if (record == null) {
             this.currentOffset = record.kafkaOffset();
         } else {
             try {
                 reentrantReadWriteLock.readLock().lock();
                 this.currentOffset = record.kafkaOffset();
-
-                fileWriter.write(value, record.kafkaOffset());
+                fileWriter.writeData(record,record.kafkaOffset());
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java:112]:
```suggestion
                fileWriter.writeData(record, record.kafkaOffset());
```

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/AvroRecordWriterProvider.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+      @Override
+      public void write(SinkRecord record) throws IOException {
+        if (schema == null) {
+          schema = record.valueSchema();
+          try {
+            log.info("Opening record writer for: {}", filename);
+            org.apache.avro.Schema avroSchema = avroData.fromConnectSchema(schema);
+            writer.setFlushOnEveryBlock(true);
+            writer.setCodec(CodecFactory.fromString(conf.getAvroCodec()));
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/AvroRecordWriterProvider.java:40]: we can comment the codec line for now.

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java [hunk 5]
```diff
@@ -19,26 +19,26 @@
...
+    private String appKey = System.getProperty("appKey","xxx");
+    private String authority = System.getProperty("authority","xxxx");
+    private String cluster = System.getProperty("cluster","xxxx");
+    private String database = System.getProperty("database","xxxx");
     private String tableBaseName = System.getProperty("table", testPrefix + UUID.randomUUID().toString().replace('-', '_'));
     private String basePath = Paths.get("src/test/resources/", "testE2E").toString();
     private Logger log = Logger.getLogger(this.getClass().getName());
 
     @Test
-//    @Ignore
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java:39]: we should ignore these tests.

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/integeration/AzureKustoConnectorIT.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.jupiter.api.Assertions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.*;
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/integeration/AzureKustoConnectorIT.java:25]: unroll this import.
Also, remove these tests for now. They will go in a separate PR.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java [hunk 7]
```diff
@@ -222,6 +214,45 @@ void flushByTimeImpl() {
...
+        currentFile.numRecords++;
+        if (this.flushInterval == 0 || currentFile.rawBytes > fileThreshold) {
+            rotate(offset);
+            resetFlushTimer(true);
+        }
+    }
+
+    public void initializeRecordWriter(SinkRecord record) {
+        if (record.value() instanceof Map) {
+            recordWriterProvider = new JsonRecordWriterProvider();
```
[the remark this answers is not in the document]
Comment (anmolbansal144) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:243]: This will initialize only one time when record writer is null https://github.com/hashedin/kafka-sink-azure-kusto/pull/8/files#diff-4031a74f967616255a0236551b9526baR222

File: pom.xml [hunk 8]
```diff
@@ -69,6 +69,7 @@
     </developers>
     <properties>
         <kafka.version>1.0.0</kafka.version>
+        <kafka.scala.version>2.12</kafka.scala.version>
```
Comment (fahadhasher) [inline pom.xml:72]: I guess this is not required since it's not being referenced anywhere.

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fahadhasher) [reply pom.xml:149]: Shall we remove dependencies related to integration tests?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/AvroRecordWriterProvider.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          // NonRecordContainers to just their value to properly handle these types
+          if (value instanceof NonRecordContainer) {
+            writer.append(((NonRecordContainer) value).getValue());
+          } else {
+            writer.append(value);
+          }
+      }
+
+      @Override
+      public void close() {
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/AvroRecordWriterProvider.java:58]: are closing the writer on task.close()??

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriterTest.java [hunk 10]
```diff
@@ -143,18 +153,18 @@ public void testGzipFileWriterFlush() throws IOException, InterruptedException {
...
+        SinkRecord record1 = new SinkRecord("topic", 1, null, null, Schema.BYTES_SCHEMA, msg2.getBytes(), 10);
+        fileWriter2.writeData(record1, null);
         Thread.sleep(1010);
 
         Assert.assertEquals(files.size(), 2);
 
         List<Long> sortedFiles = new ArrayList<>(files.values());
         sortedFiles.sort((Long x, Long y) -> (int) (y - x));
-        Assert.assertEquals(sortedFiles, Arrays.asList((long) 14, (long) 7));
+        Assert.assertEquals(sortedFiles, Arrays.asList((long) 15, (long) 8));
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriterTest.java:167]: why are we changing the existing asserts?

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java [hunk 11]
```diff
@@ -67,12 +68,12 @@ public void testE2ECsv() throws URISyntaxException, DataClientException, DataSer
...
             props.ingestionProperties.setIngestionMapping("mappy", IngestionMapping.IngestionMappingKind.Csv);
 
-            TopicPartitionWriter writer = new TopicPartitionWriter(tp, ingestClient, props, Paths.get(basePath, "csv").toString(), fileThreshold, flushInterval);
+            TopicPartitionWriter writer = new TopicPartitionWriter(tp, ingestClient, props, Paths.get(basePath, "csv").toString(), fileThreshold, flushInterval,config);
             writer.open();
 
             List<SinkRecord> records = new ArrayList<SinkRecord>();
             records.add(new SinkRecord(tp.topic(), tp.partition(), null, null, Schema.BYTES_SCHEMA, messages[0].getBytes(), 10));
-            records.add(new SinkRecord(tp.topic(), tp.partition(), null, null, null, messages[0], 10));
+            records.add(new SinkRecord(tp.topic(), tp.partition(), null, null, null, messages[0].getBytes(), 10));
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java:76]: why this change?

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriterTest.java [hunk 12]
```diff
@@ -1,15 +1,17 @@
...
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.testng.Assert;
 
 import java.io.*;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Paths;
 import java.time.Instant;
 import java.util.*;
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriterTest.java:17]: unroll these wildcard imports.

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTaskTest.java [hunk 13]
```diff
@@ -113,9 +114,11 @@ public void testSinkTaskPutRecordMissingPartition() throws Exception {
     }
 
     @Test
+    @Ignore
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTaskTest.java:117]: we should only ignore E2E tests.

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/JsonRecordWriterProviderTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      i++;
+    }
+    assertEquals(rd.getDataSize(),100);
+    file.delete();
+  }
+
+  protected Map<String, String> getProperties() {
+    Map<String, String> props = new HashMap<>();
+    props.put("connector.class", "com.microsoft.azure.kusto.kafka.connect.sink.KustoSinkConnector");
+    props.put("bootstrap.servers", "localhost:9092");
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/formatWriter/JsonRecordWriterProviderTest.java:50]: keep only the necessary configs.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 15]
```diff
@@ -121,10 +120,12 @@ public static ConfigDef getConfig() {
       final String connectionGroupName = "Connection";
       final String writeGroupName = "Writes";
       final String errorAndRetriesGroupName = "Error Handling and Retries";
+      final String formatGroupName = "formatter";
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:123]: not required

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java [hunk 16]
```diff
@@ -121,10 +120,12 @@ public static ConfigDef getConfig() {
...
 
       int connectionGroupOrder = 0;
       int writeGroupOrder = 0;
       int errorAndRetriesGroupOrder = 0;
+      int formatGroupNameOrder =0;
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkConfig.java:128]: not required

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java [hunk 17]
```diff
@@ -48,14 +49,14 @@
         this.eventDataCompression = ingestionProps.eventDataCompression;
         this.currentRetries = defaultRetriesCount;
         this.reentrantReadWriteLock = new ReentrantReadWriteLock(true);
+        this.config=config;
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java:52]:
```suggestion
        this.config = config;
```

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java [hunk 18]
```diff
@@ -150,9 +133,11 @@ void open() {
                 fileThreshold,
                 this::handleRollFile,
                 this::getFilePath,
-                !shouldCompressData ? 0 : flushInterval,
+                flushInterval,
                 shouldCompressData,
-                reentrantReadWriteLock);
+                reentrantReadWriteLock,
+                config,
+            ingestionProps);
```
Comment (fahadhasher) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java:140]:
```suggestion
                ingestionProps);
```

File: src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java [hunk 19]
```diff
@@ -19,26 +19,27 @@
 import java.io.IOException;
 import java.net.URISyntaxException;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
+import java.util.*;
```
Comment (fahadhasher) [inline src/test/java/com/microsoft/azure/kusto/kafka/connect/sink/E2ETest.java:22]: unroll these imports.
